### PR TITLE
OpenAPI/Swagger JSON type schema support + many small fixes I came across

### PR DESCRIPTION
**Pull request by Jolanrensen**
As described by https://github.com/Kotlin/dataframe/issues/142

This PR contains a custom OpenAPI -> DF Marker conversion and implementation in the Gradle- and KSP plugin.
There are also a lot of tests, but I'm not yet 100% confident I've caught every edge case (there are many...).
Docs are updated too.
Let me know if there are any major things (or minor ones, all are welcome) that need changing or if you have some more testing ideas.

### CODE REVIEW

File: plugins/symbol-processor/src/test/kotlin/org/jetbrains/dataframe/ksp/DataFrameSymbolProcessorTest.kt [hunk 1]
```diff
@@ -888,6 +1043,173 @@ class DataFrameSymbolProcessorTest {
         }
     }
 
+    private val petstoreYaml = File("../../core/src/test/resources/petstore.yaml")
+    private val petstoreYamlUrl = "https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.yaml"
```
Comment (koperagen) [inline plugins/symbol-processor/src/test/kotlin/org/jetbrains/dataframe/ksp/DataFrameSymbolProcessorTest.kt:1047]: is there a reason why this is downloaded directly from their github instead of using local copy + useHostedFile?

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt [hunk 2]
```diff
@@ -0,0 +1,1218 @@
...
+import org.jetbrains.kotlinx.dataframe.codeGen.ValidFieldName
+import org.jetbrains.kotlinx.dataframe.codeGen.plus
+import org.jetbrains.kotlinx.dataframe.schema.ColumnSchema
+import java.io.File
+import java.io.InputStream
+import java.net.URL
+import kotlin.reflect.KType
+import kotlin.reflect.typeOf
+
+public fun main() {
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:38]: also look like leftovers

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt [hunk 3]
```diff
@@ -0,0 +1,1218 @@
...
+    }
+}
+
+/**
+ * Converts named OpenApi schemas to a list of [MyMarker]s.
+ * Will cause an exception for circular references, however they shouldn't occur in OpenApi specs.
+ */
+private fun Map<String, Schema<*>>.toMarkers(): List<MyMarker> {
+    // Convert the schemas to toMarker calls that can be repeated to resolve references.
+    val retrievableMarkers = mapValues { (typeName, value) ->
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:334]: Can you explain this part with references, resolve, retrieve and  ProduceAdditionalMarker? It seems i don't know enough about open api to understand it

**Review by koperagen — DISMISSED** [review]

File: examples/idea-examples/movies/build.gradle.kts [hunk 4]
```diff
@@ -22,3 +22,10 @@ tasks.withType<KotlinCompile> {
     kotlinOptions.jvmTarget = "1.8"
 }
 
+//dataframes {
```
[the remark this answers is not in the document]
Comment (Jolanrensen) [reply examples/idea-examples/movies/build.gradle.kts:25]: nope XD thanks!

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt [hunk 5]
```diff
@@ -17,17 +18,27 @@ import java.util.ServiceLoader
...
 
     // `DataFrame.Companion.read` methods uses this to sort list of all supported formats in ascending order (-1, 2, 10)
     // sorted list is used to test if any format can read given input
     public val testOrder: Int
 
-    public fun createDefaultReadMethod(pathRepresentation: String?): DefaultReadDfMethod
+    public fun createDefaultReadMethod(pathRepresentation: String? = null): DefaultReadDfMethod
+}
+
+public interface SupportedDataFrameFormat : SupportedFormat {
```
[the remark this answers is not in the document]
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt:31]: cool! didn't know that

File: plugins/symbol-processor/src/test/kotlin/org/jetbrains/dataframe/ksp/DataFrameSymbolProcessorTest.kt (hunk 1 above)
Comment (Jolanrensen) [reply plugins/symbol-processor/src/test/kotlin/org/jetbrains/dataframe/ksp/DataFrameSymbolProcessorTest.kt:1047]: I wanted to test whether that would work too, but local might be better

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt (hunk 3 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:336]: This is actually something I came up with myself. OpenApi provides schemas for all of the types used. For each type, we want to generate a Marker (Which can be an interface, enum, etc.). However, the OpenApi schema is not ordered per se, so when we are reading the schema it might be that we have a reference to a (super)type (which are queried using `getRefMarker`) for which we have not yet created a Marker. In that case, I "pause" that one (by returning CannotFindRefMarker) and try to read another type schema first. Circular references cannot exist since it's encoded in JSON, so we never get stuck in an infinite loop. When all markers are "retrieved" (so turned from a RetrievableMarker to a MarkerResult.Success), we're done and have converted everything!

As for `produceAdditionalMarker`: In OpenAPI not all enums/objects have to be defined as a separate schema. Although recommended, you can still define an object anonymously directly as a type. For this, I have `produceAdditionalMarker` since during the conversion of a schema->Marker we get an additional new Marker.

**Comment by Jolanrensen** [comment]
okay, I looked over additionalProperties... https://swagger.io/docs/specification/data-models/dictionaries/
Let's see how I can add that

**Comment by Jolanrensen** [comment]
Since I changed so much I'm gonna provide a small overview of the changes I made. Might make it easier to review. I'll go through the changed files:

- Linting changes (sorry, but improves readability)
- Docs for OpenAPI and jsonoptions
- replaced all occurrences of "splitted" with "split" because English
- updated kotlinDatetime and made it api() for jupyter etc.
- ImportDataSchema gained jsonOptions with typeClashTactic and keyValuePaths
- starting with kdoc in some functions (more will come later)
- convertTo
  - `convertIf` in dsl: way more powerful than `convert(KType, KType)`, allows you to specify via a condition function whether you want to do a certain conversion and provides fromType and toSchema in ConverterScope. `convert` still has priority over `convertIf`
  - convertTo can now happen for any empty df (both 0 rows or 0 columns)
  - manual converters can now happen for any type of target column, not just value columns. user conversion happens before df conversion making some conversions that were previously impossible possible.
  - Value columns of datarows can become column groups, same as value columns with all nulls.
  - value columns of dataframes can become frame columns
  - absent columns can be created iff it's a nullable typed value column, DataRow<Something?> for a group column or frame column.
- containsNoData() helper function (created that once but don't use it anymore, might still be useful? opinion needed)
- enums can implement `DataSchemaEnum` to control how they are (en/de)coded from/to datasets instead of using just their name
- code generation can now generate enums and type aliases too
- `DefaultReadDfMethod` now provides actual `Marker` instead of just a name
- nullability helpers for FieldType
- bugfix for `MarkersExtractor` regarding `nullableProperties`
- `contains` in `BaseColumn` is now `operator fun`
- Bugfix for concat of DataFrames without columns (rows must still count up, else schema conversions will break)
- String+Number can become Comparable, this was always the case but order dependent. That now always gives the same result thanks to a bugfix in `commonParents()` and `commonType()`
- ISO_DATE_TIME support in parser from string
- `CodeGenerator.Companion.urlReader` is now split into `urlDfReader` and `urlCodeGenReader` where the former is used to generate a dataframe from the data and types from that, while the latter directly generates types (for openapi)
- Similar split for `SupportedFormat` into `SupportedDataFrameFormat` and `SupportedCodeGenerationFormat`
- `createColumn()`
  - bugfix for empty iterables always making column groups. Now uses guesstype or creates value column
  - bugfix for iterables with just nulls always making a frame column, now also uses guesstype or creates value column
  - can now create Column group with iterable of datarows too
- small fix in printing newlines of data schemas
-  `createEmptyColumn` and `createEmptyDataFrame` variants with `numberOfRows`
- `ColumnSchema`, aside from `type` now also has `contentType` for extra type info for Group- and Frame columns (instead of just getting `DataRow<*>` or `DataFrame<*>`. Very useful for new `convertIf` method. `intersectSchema` also tries to merge these `contentType`s if possible. `extractSchema` will make them `Any?`.
- `toSnakeCase()` helper function
- TODO more...

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt [hunk 6]
```diff
@@ -164,20 +175,30 @@ internal fun commonParents(classes: Iterable<KClass<*>>): List<KClass<*>> =
...
-    }
+    }.sortedBy { it.simpleName } // make sure the order is stable to avoid bugs
 
-internal fun baseType(types: Set<KType>): KType {
+/**
+ * Returns the common type of the given [types].
+ *
+ * @param types the types to find the common type for
+ * @param listifyValues if true, then values and nulls will be wrapped in a list if they appear among other lists.
+ *   For example: `[Int, Any?, List<Int>]` will become `List<Int>` instead of `Any?`. If there is another collection
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt:198]: is that so? 
It seems like a bug. Because type of null should be Nothing?

File: core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/groupBy.kt [hunk 7]
```diff
@@ -9,27 +9,42 @@ class GroupByTests {
 
     @Test
     fun `groupBy values with nulls`() {
-        val df = dataFrameOf("a", "b")(
+        val df = dataFrameOf(
```
Comment (koperagen) [inline core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/groupBy.kt:12]: i don't think this file should be reformatted in this commit / PR. It makes it harder to see actual changes. i mean, there are none, right?

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt [hunk 8]
```diff
@@ -62,70 +64,82 @@ internal interface TypeRenderingStrategy {
...
-    override fun BaseField.renderFieldType(): Code =
+    override fun BaseField.renderFieldType(asAccessor: Boolean): Code =
         when (val fieldType = fieldType) {
-            is FieldType.ValueFieldType -> fieldType.typeFqName.shorten()
-            is FieldType.GroupFieldType -> "${DataRow::class.simpleName}<${fieldType.markerName}>"
+            is FieldType.ValueFieldType ->
+                fieldType.typeFqName.shorten()
+
+            is FieldType.GroupFieldType ->
+                if (asAccessor) "$DataRow<${fieldType.markerName}>"
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt:144]: Similar conversion happens for FrameFieldType and List. You can generate List<fieldType.markerName> for usages inside data schema declaration and DataFrame<>... for usages in accessors. What do you think about changing this argument to an enum UseSite { DataSchemaDeclaration, GeneratedAccessor }? Or even splitting in two functions. Took me quite a while to understand what's this about

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt [hunk 9]
```diff
@@ -0,0 +1,1464 @@
...
+    // convert DataRows to Maps if required by the schema
+    convert(
+        from = { it == typeOf<DataRow<*>>() },
+        to = { // any type of Map<String, _> or Map<String, _>?
+            (it.isSubtypeOf(typeOf<Map<*, *>>()) || it.isSubtypeOf(typeOf<Map<*, *>?>())) &&
+                it.arguments.getOrNull(0)?.type == typeOf<String>()
+        }
+    ) { (it as DataRow<*>).toMap() }
+
+    // convert DataFrame to DataFrame<Any?> if required by the schema
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt]: What's the special meaning of DataFrame<Any?>? 
Previously, i think, you could just take any existing dataframe would be a valid value of type DataFrame<Any?>. But here's converting

Comment (koperagen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt]: ok, it got commented out :D

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt (hunk 6 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt:198]: Definitely not `Nothing?` right? That only appears for thrown errors and return statements. `Nothing?` cannot have an instance. `Int, Any?, List<Int>` will become `Any?` as far as I know.

File: core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/groupBy.kt (hunk 7 above)
Comment (Jolanrensen) [reply core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/groupBy.kt:12]: Unfortunately I have many of these reformats :/ Also for files I've touched but reverted in later commits. So there's no beginning trying to revert them all again I think. We could also view them as one of the "small fixes" part of the PR if that's okay with you :)

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt (hunk 9 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt]: This is not present anymore in later commits :)

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt (hunk 6 above)
Comment (koperagen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt:198]: Nothing appears for thrown errors, but `Nothing?` is a special type for null values 
image

File: core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/groupBy.kt (hunk 7 above)
Comment (koperagen) [reply core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/groupBy.kt:12]: Sure, let's keep these changes. Just saying for the future

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt (hunk 6 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt:198]: Oh, I see now. I made an error in the KDoc. `[Int, Any?, List<Int>]` becomes `List<Any?>?`.
`[Int, Nothing?, List<Int>]` will become `List<Int>` (or at least, it should, but it doesn't atm... Bug :)  )

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt (hunk 8 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt:144]: 2 functions instead of 1 boolean argument is always better indeed :) 
Where is this conversion in lists you talk about?

Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt]: Okay, check out the latest commit, there I split the methods, renamed a bit and added explanation+examples :) should be a lot clearer what's happening now!

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/ReplCodeGeneratorImpl.kt [hunk 10]
```diff
@@ -68,7 +68,7 @@ internal class ReplCodeGeneratorImpl : ReplCodeGenerator {
                         // property scheme is valid for current data frame, but we should also check that all compatible open markers are implemented by it
                         val requiredBaseMarkers =
                             getRequiredMarkers(columnSchema, registeredMarkers.values)
-                        if (requiredBaseMarkers.all { currentMarker.implements(it) }) {
+                        if (requiredBaseMarkers.any() && requiredBaseMarkers.all { currentMarker.implements(it) }) {
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/ReplCodeGeneratorImpl.kt:71]: is this a fix for some bug? would be nice to have a test to see why this code was changed (or maybe there is? i didn't find it)

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt (hunk 8 above)
Comment (koperagen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt]: I mean
@DataSchema
interface Name {
    val firstName: String
}   

@DataSchema
interface Departments {
    val people: List<Name> // Name is DataSchema, so accessor field type is DataFrame. 
    // same as
    val people: DataFrame<Name>
}

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/ReplCodeGeneratorImpl.kt (hunk 10 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/ReplCodeGeneratorImpl.kt:71]: yup. Accessors wouldn't be generated for my interface-nested DataSchemas without this fix. I think it was just a misinterpretation of `.all {}` again, that's why I didn't write an explicit test here, but the OpenApi Jupyter test breaks if this `.any()` check isn't there.

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt [hunk 11]
```diff
@@ -16,18 +19,47 @@ import java.net.URL
...
+ * Note, this doesn't add functionality to [DataFrame.Companion.read], just [ImportDataSchema] and Gradle plugin.
+ *
+ * Return type will be a [CodeWithConverter] for which [CodeWithConverter.converter] needs to be called
+ * to get the generated code with the correct top-level interface name.
+ */
+public interface SupportedCodeGenerationFormat : SupportedFormat {
+
+    public fun readCodeForGeneration(
+        stream: InputStream,
+        generateHelperCompanionObject: Boolean = false,
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt:63]: this in fact needed to support val ApiGuru = importDataSchema() notation in Jupyter, right? I think it's worth mentioning

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/KeyValueProperty.kt [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package org.jetbrains.kotlinx.dataframe.api
+
+import org.jetbrains.kotlinx.dataframe.annotations.ColumnName
+import org.jetbrains.kotlinx.dataframe.annotations.DataSchema
+
+/** A [DataSchema] interface / class can implement this if it represents a map-like data schema (so key: value). */
+public interface KeyValueProperty<T> {
+    // needs to be explicitly overridden in @DataSchema interface, otherwise extension functions won't generate (TODO)
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/KeyValueProperty.kt:8]: this interface can be annotated with @DataSchema itself. then extensions going to be a part of the library API and could be called on every marker that implements this interface

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convert.kt [hunk 13]
```diff
@@ -76,41 +77,49 @@ internal fun AnyCol.convertToTypeImpl(to: KType): AnyCol {
...
-                    DataColumn.createValueColumn(name, values, to.withNullability(nullsFound))
-                }
-                else -> throw TypeConverterNotFoundException(from, to)
-            }
-            else -> {
+    if (from == to) return this
+
+    if (from.isSubtypeOf(to)) try {
+        return (this as DataColumnInternal<*>).changeType(to.withNullability(hasNulls()))
+    } catch (e: UnsupportedOperationException) { /* */
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convert.kt:88]: do you remember why?

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convertTo.kt [hunk 14]
```diff
@@ -17,106 +25,209 @@ import org.jetbrains.kotlinx.dataframe.exceptions.ExcessiveColumnsException
...
+    @Suppress("UNCHECKED_CAST")
     override fun <A, B> convert(from: KType, to: KType, converter: (A) -> B) {
-        converters[from.withNullability(false) to to.withNullability(false)] = Converter(converter as (Any?) -> Any?, !from.isMarkedNullable)
+        converters[from.withNullability(false) to to.withNullability(false)] =
+            Converter({ converter(it as A) }, !from.isMarkedNullable)
     }
 
-    fun getConverter(from: KType, to: KType): Converter? {
-        return converters[from.withNullability(false) to to.withNullability(false)]
+    override fun convertIf(
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convertTo.kt:53]: Do you have some representative example for this API? Honestly, i can't easily understand recursive list thing :(

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt [hunk 15]
```diff
@@ -0,0 +1,2173 @@
...
+
+    if (file.extension.lowercase() != "json") {
+        return false
+    }
+
+    return isOpenApiStr(file.readText())
+}
+
+/** Parse and read OpenApi specification to [DataSchema] interfaces. */
+public fun readOpenApi(
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:345]: why not make DataFrame.Companion receiver?

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt [hunk 16]
```diff
@@ -0,0 +1,2173 @@
...
+         * DataFrame. This is to prevent `gather { all() }` from creating a lot of `null` values due to
+         * merging of ColumnGroups.
+         *
+         *
+         * @receiver the DataFrame to convert to an [AdditionalProperty] [DataFrame].
+         * @param convertTo optional [ConvertSchemaDsl] to specify extra conversions that need to occur.
+         * @return a [DataFrame] of [AdditionalProperty]s.
+         */
+        @Suppress("UNCHECKED_CAST")
+        public fun AnyFrame.convertToAdditionalProperties(
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:531]: maybe mention that this used in generated code?

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt [hunk 17]
```diff
@@ -0,0 +1,2173 @@
...
+import java.io.File
+import java.io.InputStream
+import java.net.URL
+import kotlin.String
+import kotlin.reflect.KType
+import kotlin.reflect.full.hasAnnotation
+import kotlin.reflect.full.isSubtypeOf
+import kotlin.reflect.jvm.jvmErasure
+import kotlin.reflect.typeOf
+
```
Comment (koperagen) [inline core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:75]: wdyt about splitting this file? also, at some point we started to put formats in different modules

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt (hunk 6 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt]: All handled in latest commit :)

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt (hunk 8 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/codeGen/CodeGeneratorImpl.kt]: Hmm, I think this is handled on the schema generation side already. `List<Name>` is a `ValueColumn` type. Can't really seem to find where the distinction happens with wrapping `DataRow` around it or not

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt (hunk 11 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt:63]: yup!

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/KeyValueProperty.kt (hunk 12 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/KeyValueProperty.kt:8]: sure, why not :)

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convert.kt (hunk 13 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convert.kt:88]: Yes, I believe this fixed a bug where converting from ColumnGroupImpl or FrameColumnImpl to something would crash because it doesn't have the `changeType` option. However, `convertTo` can now support conversions for any type of column, so the exception is caught here to give any user-given conversions a try first, otherwise, a TypeConverterNotFoundException is thrown.

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convertTo.kt (hunk 14 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/convertTo.kt:53]: For `convertIf`? Sure! It's not that difficult. `convert.with` can only support exact type matches, which can be a problem when you cannot define exact types for everything you want to catch. Let's say you want to write a converter that would convert all numbers to strings:
```kotlin
convert<Number>.with { it.toString() }
```
This wouldn't work, since `typeOf<Double>() != typeOf<Number>()`. That's where `convertIf` can come in handy. With this function you can manufacture any type of converter you like and you can use the full capabilities of Kotlin reflect you like:
```kotlin
convertIf({ fromType, toSchema -> 
    fromType.withNullability(false).isSubtypeOf(typeOf<Number>()) &&
        toSchema.type.withNullability(false) == typeOf<String>()
}) {
    it.toString()
}
```

For OpenApi I needed to provide a converter for `DataRow -> Any`, however, these converters only work on top-level types, so if a `DataRow->Any` conversion needs to happen in a `List<List<List<DataFrame<*>>>`, then we need my recursive function which makes `convertIf` accept any of these types and convert it accordingly.

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt (hunk 15 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:345]: Because you don't create a `DataFrame` ;)

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/DataRowApi.kt [hunk 18]
```diff
@@ -22,6 +23,24 @@ import kotlin.reflect.KType
 public fun AnyRow.isEmpty(): Boolean = owner.columns().all { it[index] == null }
 public fun AnyRow.isNotEmpty(): Boolean = !isEmpty()
 
+/** Returns `true` even if the columns contain empty iterables, empty dataframes etc. */
+public fun Any?.containsNoData(): Boolean =
```
[the remark this answers is not in the document]
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/DataRowApi.kt:27]: Okay I agree :) will remove it

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt (hunk 16 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:531]: sure

File: core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt (hunk 17 above)
Comment (Jolanrensen) [reply core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/openapi.kt:75]: Okay splitting off openapi works up 'till the part where I decide to read a file as json or as openapi. Need to think of something clever or universal to make that work elegantly :)
